### PR TITLE
Update plugins/data from Github

### PR DESCRIPTION
**Pull request by etp12**
# Plugin Downloader

This pull request adds the ability for users to update plugins directly from its Github repository. A python script goes uses the Github API to gather information from these repositories and produces a .json file with the needed information here: [https://avogadro.cc/plugins.json](https://avogadro.cc/plugins.json).

This extension accesses this json file and displays a table of the plugins with their name, description, and current release/time of last update and allows them to choose which ones they want to update. Clicking on a repository in the table displays the README.md for the repository. 

When the user clicks download the extension downloads a zip of each selected repository one at a time to "QStandardPaths::AppDataLocation" which should be "C:/Users/<USER>/AppData/Roaming/<APPNAME>" on Windows or "~/Library/Application Support/<APPNAME>" on macOS. The archive is then unzipped with Libarchive, into a subdirectory based on the plugin's type. 

This is what the ui currently looks like with one of the repo's README being displayed:

### CODE REVIEW

**Comment by ghutchis** [comment]
Could you post a screenshot or two? Thanks. :-)

**Comment by etp12** [comment]
One note, last PR there was an issue with my commit authorship and it looks like it's still present in some of these commits. I had set my username on the command line to "etp12" instead of "Ethan Pavolik" like it is on the Github website. This should be fixed for all my future commits.

File: avogadro/qtplugins/plugindownloader/downloaderwidget.cpp [hunk 1]
```diff
@@ -0,0 +1,242 @@
+#include "downloaderwidget.h"
```
Comment (cryos) [inline avogadro/qtplugins/plugindownloader/downloaderwidget.cpp]: We need copyright headers in the new files.

File: avogadro/qtplugins/plugindownloader/downloaderwidget.cpp [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+#include "zipextracter.h"
+#include "ui_downloaderwidget.h"
+
+namespace Avogadro {
+namespace QtPlugins {
+
+DownloaderWidget::DownloaderWidget(QWidget *parent)
+    : QDialog(parent), ui(new Ui::DownloaderWidget) {
```
Comment (cryos) [inline avogadro/qtplugins/plugindownloader/downloaderwidget.cpp]: We can probably just fix this with clang-format, but functions etc have the opening { on their own line

File: avogadro/qtplugins/plugindownloader/downloaderwidget.h [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+namespace Ui {
+class DownloaderWidget;
+}
+
+namespace Avogadro {
+
+namespace QtPlugins {
+
+class DownloaderWidget : public QDialog {
```
Comment (cryos) [inline avogadro/qtplugins/plugindownloader/downloaderwidget.h]: Same with class declarations, opening { should be on a newline.

File: avogadro/qtplugins/plugindownloader/downloaderwidget.h [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+  void updateRepoData();
+  void getCheckedRepos();
+  void handleRedirect();
+  void unzipPlugin();
+
+ private:
+  typedef struct repo {
+    QString name;
+    QString description;
+    QString release_version;
```
Comment (cryos) [inline avogadro/qtplugins/plugindownloader/downloaderwidget.h]: We also prefer camel case to underscores

File: avogadro/qtplugins/plugindownloader/downloaderwidget.h [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    QString url;
+    QString name;
+    QString type;
+  };
+  void downloadNextPlugin();
+  void getRepoData();
+  void downloadNext();
+  bool checkSHA1(QByteArray);
+
+  struct repo *repoList;
```
Comment (cryos) [inline avogadro/qtplugins/plugindownloader/downloaderwidget.h]: Member variables should all use an m_ prefix to show that they are member variables.

File: avogadro/qtplugins/plugindownloader/downloaderwidget.h [hunk 6]
```diff
@@ -0,0 +1,83 @@
+#ifndef AVOGADRO_DOWNLOADERWIDGET_H
+#define AVOGADRO_DOWNLOADERWIDGET_H
+
+#include <QtNetwork/QNetworkAccessManager>
```
Comment (cryos) [inline avogadro/qtplugins/plugindownloader/downloaderwidget.h]: It is good practice to minimize what is included in the header, predeclaring where reasonable, this doesn't matter as much in plugins as they are not reused but it is worth bearing in mind.

File: avogadro/qtplugins/plugindownloader/downloaderwidget.h [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+}
+
+namespace Avogadro {
+
+namespace QtPlugins {
+
+class DownloaderWidget : public QDialog {
+  Q_OBJECT
+
+ public:
```
Comment (cryos) [inline avogadro/qtplugins/plugindownloader/downloaderwidget.h]: public: and other sections have no spaces preceding them.

File: avogadro/qtplugins/plugindownloader/plugindownloader.h [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+
+/**
+ * @brief Downlaods Github repos and extracts their contents into a Avogadro
+ * folder for plugins, molecule data, etc..
+ */
+
+class PluginDownloader : public QtGui::ExtensionPlugin {
+  Q_OBJECT
+
+ public:
```
Comment (cryos) [inline avogadro/qtplugins/plugindownloader/plugindownloader.h]: Same here, no indent for public: etc, and opening { for classes, functions etc on their own line.

File: avogadro/qtplugins/plugindownloader/zipextracter.h [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+namespace Avogadro {
+
+namespace QtPlugins {
+
+class ZipExtracter {
+ public:
+  ZipExtracter();
+  ~ZipExtracter();
+  char *convert(const std::string &);
+  int copy_data(struct archive *ar, struct archive *aw);
```
Comment (cryos) [inline avogadro/qtplugins/plugindownloader/zipextracter.h]: Prefer camel case copyData over copy_data

**Comment by etp12** [comment]
@cryos These should all be fixed now. I used the copyright header I found in some other files and moved a lot of the includes out of the header files.

**Comment by cryos** [comment]
I edited the branch to fix authorship, and remove all of the merge commits. This is now based on master, and almost works after the commit I added on the end. For some reason I am seeing a symbol resolution error, and looking into what might be wrong there.

**Comment by cryos** [comment]
I have this building and linking, but am left with some usability questions. It looks like the dialog is not correctly parented - when I close the main Avogadro window the download dialog remains open.

What is "Download All" intended to do? Does it download the items I select, everything in teh list? In my local testing whether I selected things or not it didn't seem to download anything (or offer me indications that it was downloading items, there was an error, etc).

**Comment by etp12** [comment]
"Download All" should only download the items you have selected, I agree that should be changed to "Download" or "Download Selected". To convey messages to the user of the progress of the download, would using the TextArea on the right work? I could have it show whether the download/extraction was successful and possibly show the download location or any errors that were encountered. I'll also fix the parenting issue.

For the actual downloading, the docs say that QStandardPaths::AppDataLocation should resolve to "~/.local/share/<APPNAME>" in Linux. After selecting plugins and clicking "Download All" are the extracted files showing in this directory, or is an /Avogadro folder being created at all here?

**Comment by cryos** [comment]
ls ~/.local/share/OpenChemistry/Avogadro/ -lthtotal 1.1M
-rw-r--r-- 1 marcus users 466K Mar 12 15:48 molecules.zip
-rw-r--r-- 1 marcus users 599K Mar 12 15:48 crystals.zip
drwxr-xr-x 5 marcus users 4.0K Mar 12 15:47 other
-rw-r--r-- 1 marcus users  22K Mar 12 15:42 avogenerators.zip
drwx------ 3 marcus users 4.0K Dec  5 11:07 QtWebEngine

It wasn't obvious anything had downloaded, I don't see anything extracted so maybe that failed, but no feedback either graphically or in the console either way. I didn't see a way to use these things, I think we want to confirm that things downloaded, and also confirm they were extracted/will be used.

**Comment by etp12** [comment]
Originally we had planned to extract each plugin into a directory based on it's type - whether it was just data, tool, etc. This information has yet to be added to the master "plugins.json" hosted [here](https://avogadro.cc/plugins.json), so in the meantime all plugins were being extracted to a default "other" directory when the plugin's type was not in the master "plugins.json". Once this is added they should appear in the relevant directory. 

After talking to Geoff today, we also agreed that the downloaded .zip files should be deleted after unzipping to avoid confusion. This, along with the other issues (UI, window parenting, and printing download info to the screen) should be fixed now.

File: CMakeLists.txt [hunk 10]
```diff
@@ -45,6 +45,7 @@ option(USE_OPENGL "Enable libraries that use OpenGL" ON)
 option(USE_HDF5 "Enable optional HDF5 features" OFF)
 option(USE_QT "Enable libraries that use Qt 5" ON)
 option(USE_VTK "Enable libraries that use VTK" OFF)
+option(USE_LIBARCHIVE "Enable optional Libarchive features" ON)
```
Comment (cryos) [inline CMakeLists.txt]: I don't see how the use of libarchive is actually optional at this point, I think we should make it optional.

File: avogadro/qtplugins/CMakeLists.txt [hunk 11]
```diff
@@ -113,6 +113,7 @@ add_subdirectory(networkdatabases)
 add_subdirectory(openbabel)
 add_subdirectory(quantumoutput)
 add_subdirectory(playertool)
```
Comment (cryos) [inline avogadro/qtplugins/CMakeLists.txt]: if(USE_LIBARCHIVE) here would do it, I can add it in.

**Pull request merged**